### PR TITLE
Fix: substitution of binding guards does not crash if splits present

### LINKED ISSUE #2748: Incorrect handling of slit quantifier expressions in substitution of binding guards
Reproduction:

```
function method f(x: int): int { 10 - x * x }

function method BindingGuardTest(): int {
  var x: nat := 1;
  assert true by {
    if i :| 0 <= i < 10 && (f(i) == f(i+1) || f(i) == f(i+2)) {
    }
  }
  2
}
```

Error:
``
Unhandled exception. System.AggregateException: One or more errors occurred. (Unable to cast object of type 'Microsoft.Dafny.BinaryExpr' to type 'Microsoft.Dafny.ComprehensionExpr'.)
 ---> System.InvalidCastException: Unable to cast object of type 'Microsoft.Dafny.BinaryExpr' to type 'Microsoft.Dafny.ComprehensionExpr'.
   at Microsoft.Dafny.Substituter.SubstituteComprehensionExpr(ComprehensionExpr expr, Boolean forceSubstituteOfBoundVars) in /Users/trjohnb/Code/dafny/Source/Dafny/Substituter.cs:line 920
   at Microsoft.Dafny.Substituter.SubstStmt(Statement stmt) in /Users/trjohnb/Code/dafny/Source/Dafny/Substituter.cs:line 683
``

### PR DESCRIPTION
**Pull request by MikaelMayer**
This PR fixes #2748
I think it's a correct solution, at least locally, because there was previously a recursion which would just replace the comprehension by its split quantifier equivalent. That was obviously wrong for quantifiers which were split into a binary expression
Since it wanted to do it on an equivalent sub-comprehension, I removed the recursive call and it should have the same effect. At least, the test now passes.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Substituter.cs [hunk 1]
```diff
@@ -934,8 +934,6 @@ public class Substituter {
       if (q != null && q.SplitQuantifier != null) {
         if (forceSubstituteOfBoundVars) {
           return Substitute(q.SplitQuantifierExpression);
-        } else {
```
[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:937]: Correct, but to avoid any doubts about that, note it was also the case before, as for "substituting" we were essentially returning the `.SplitQuantifierExpression`, whose `.SplitQuantifierExpression` is itself null.

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Substituter.cs:937]: Sure, but that does seem quite different, because now the translation to `SplitQuantifierExpression` is lost. Are you sure we're not missing a test-case? Could we come up with a proof that requires the quantifier expression to be split and that is substituted?

[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:937]: I'm sure of one thing, this code was wrong, because it assumed that the `.SplitQuantifierExpression` was a `ComprehensionExpr`.
Let's analyse that. Take a deep breath.

If `q.SplitQuantifierExpression` is a `ComprehensionExpr`. it means that there was no splitting at all. This is because any split would make that `q.SplitExpressions` would be a list of length >= 2 and thus the generated  `.SplitQuantifierExpression` would HAVE to be a binary expression
Hence the `q.SplitQuantifierExpression` is equal to the expression itself, minus the `.SplitQuantifier` which is now null (splitted quantifier expressions are not themselves splitted)
However, the remaining of `SubstituteComprehensionExpr` does not access `.SplitQuantifier` nor `.SplitQuantifierExpression` after line 937. Thus, accessing the `.SplitQuantifierExpression` is useless.

So, for the existing case where `e.SplitQuantifierExpression` is a `ComprehensionExpr`, the code does the same. 

Now, note that my change only changes anything when `forceSubstituteOfBoundVars` is set to false.
If you look at the three call sites, two set  `forceSubstituteOfBoundVars`  to false - binding guard if statements and binding guard if expressions, and one set `forceSubstituteOfBoundVars` to true - all the other places.

So, in essence, `forceSubstituteOfBoundVars` is set to false in this context because we are in a binding guard, and the reason why we don't want to force substituting bound variables is that these variables are used afterwards in the body of the if. By the way, when `forceSubstituteOfBoundVars == false`, the bound variables are still substituted, it's just that the current scope's variables substitutation map is updated.

In conclusion when translating a binding guard, it makes sense to preserve its atomic quantifier-ness and discard splitted expressions. After all, splitted expressions are useful primarly for proving smaller assertions, but since it's a guard here, we don't need to split, since there is nothing to prove.
The only thing that we want to prove might be the existence and unicity of such an assignment, but these assertions appear outside of a binding guard, and thus can be splitted as normal (`forceSubstituteOfBoundVars == true`)

**Comment by cpitclaudel** [comment]
I haven't followed the details of this PR, but we discussed this issue in depth with @jtristan a little while back in the comments of https://github.com/dafny-lang/dafny/pull/2745 ; is there anything helpful there?

**Comment by MikaelMayer** [comment]
> I haven't followed the details of this PR, but we discussed this issue in depth with @jtristan a little while back in the comments of #2745 ; is there anything helpful there?

Yes that was helpful. Basically, this conversation confirms my analysis above
 https://github.com/dafny-lang/dafny/pull/2745/files#r971378730 
Basically, in the PR you mentioned, it was only a code refactoring to ensure that binding guards had a different treatment, but it never addressed the issue that the split quantifier expression was not a quantifier by itself.
And your comment was right: we can just omit the second case. It was there only to ensure the previous behavior, but I agree with you it did not have any reason to stay.

File: Source/DafnyCore/Substituter.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:937]: I am not sure about your code. Why would you want the caller to be aware of the splits?

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Substituter.cs:937]: That way the binding guards are aware of discarding the splits.

[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:937]: I feel very strongly that callers handling the binding guards should not be aware of the implementation detail that consists of considering splits instead of the original expression.

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Substituter.cs:937]: > I feel very strongly that callers handling the binding guards should not be aware of the implementation detail that consists of considering splits

I think right now they already have to be, it's just not explicit. There's a comment above `SubstituteComprehensionExpr` that says `Parameter 'forceSubstituteOfBoundVars' should be set to false if and only if the expression is a binding guard`.

I would like the signature of a method to make it clear what it does, instead of having to rely on comments above the method. 

To me, the parameter name `forceSubstituteOfBoundVars` doesn't imply that anything should be done differently for splits. With the change I suggested above, I feel the method signature `SubstituteComprehensionExpr` and `SubstituteComprehensionExprDiscardingSplits` are self explanatory, while binding guards callers need to be just as aware of what they're calling as they are now, and other callers of `SubstituteComprehensionExpr` need to be less aware since now they don't need to think about what value of `forceSubstituteOfBoundVars` to pass in.

[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:937]: Callers handling the binding guard are only aware of the binding guard and that's what they tell to the substituter.
They have to tell the substituter that the bound variable in the existential expression in the guard is not contained there, its scope extend to the body of the if. And that's the only thing they are aware. They are not aware of the implementation details of the substituter at all.
To split or not to split a quantifier for verification purpose is an optimization for the verifier. This PR just says we shouldn't make this optimization for a binding guard, because otherwise it would crash. In fact, we could remove the splitting alltogether and have the roughly same verification behavior.
It's the splitting implementation which needs to be aware of guards, not the reverse way round.

[the remark this answers is not in the document]
Comment (robin-aws) [reply Source/DafnyCore/Substituter.cs:937]: I strongly agree that discarding splits is a very surprising side-effect of substitution, and I'm not really convinced by any argument that it's currently harmless. It feels like a very subtle bug waiting to happen when someone refactors code without being aware of it. I don't even think having `SubstituteComprehensionExpr` call a `SubstituteComprehensionExprDiscardingSplits` method helps that much, because anyone just looking at the name of the first isn't going to be aware of the side-effect.

Would it be tractable to add the `forceSubstituteOfBoundVars` parameter to `Substitute` with a default value, remove the incorrect cast, and just call `Substitute` on `q.SplitQuantifierExpression` instead of `SubstituteComprehensionExpr`, so that it can work as advertised on splits as well?

[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:937]: > I strongly agree that discarding splits is a very surprising side-effect of substitution

This is true only in the case of binding guards.

> Would it be tractable to add the forceSubstituteOfBoundVars parameter to Substitute with a default value, remove the incorrect cast, and just call Substitute on q.SplitQuantifierExpression instead of SubstituteComprehensionExpr, so that it can work as advertised on splits as well?

You mean, removing `SubstituteComprehensionExpr` ? No that wouldn't work for quantifiers that are in the position of a binding guard. There, we still need to update the current scope because these variables aren't scoped to the expression only, but also to the body of the if.

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Substituter.cs:937]: One that's blocking me from understanding the code well is that I don't understand the `forceSubstituteOfBoundVars` parameter. Why would you ever want to force substitution? Only doing the substitution when necessary seems strictly more performant.

[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:937]: I did more research to answer you accurately. The variable `forceSubstitutionOfBoundVars` is only used in  `CreateBoundVarSubstitutions`, and `SubstituteCasePattern` in the simple variable pattern:

https://github.com/dafny-lang/dafny/blob/74a4e2046cf6b8f053d3e7e46134fa6ec01c95cc/Source/DafnyCore/Substituter.cs#L459-L478
https://github.com/dafny-lang/dafny/blob/74a4e2046cf6b8f053d3e7e46134fa6ec01c95cc/Source/DafnyCore/Substituter.cs#L526-L544

Interestingly, in both of these cases, before `forceSubstitutionOfBoundVars`, the only test to determine if a bound variable (whether in patterns or in quantifier expressions) has to be substituted was if its type after substitution of type variables is different from its resolved type. In that case, there is a need to clone the bound variable.
And the comments make it clear that the caller is in charge of removing substitutions from the maps when the variables are out of scope.

Consider the following code:

```
`if i :| f(i) || g(i) {
  // i is accessible
 }
```
The AST of the code, modulo some flags, is
```
`if exists i :: f(i) || g(i) {
  // i is still accessible
 }
```
The splitting of the quantifier makes it that the expression is equivalent to the disjunction of the splits, i.e.
```
`if (exists i :: f(i)) || (exists i :: g(i)) {
   // which i is accessible?
 }
```
However, as you can see, the expression above (which is the split) introduces two bound variables in two different expressions. This is wrong. For comprehensions expressions, it was assumed that it never occurred until this PR, but it does. So we just can't reuse the SplitExpression here for verification purposes anyway.

When looking at the comment before `CreateBoundVarSubstitutions`:
https://github.com/dafny-lang/dafny/blob/74a4e2046cf6b8f053d3e7e46134fa6ec01c95cc/Source/DafnyCore/Substituter.cs#L919-L924
it becomes obvious that, in this context, `forceSubstitutionOfBoundVars` really means `isNotBindingGuard`, so I'm renaming this variable in this context to its positive counter-part `isBindingGuard`.
Let me know if it makes more sense and what parts need more clarification. Otherwise, we can also ask @jtristan  or @cpitclaudel  for advice because they worked on this previously.

[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:937]: `isBindingGuard` is not only used to determine if we can discard splits or not, it's also used later when substituting variables. Do you really think it would be appropriate to transform the following:

```
var newBoundVars = CreateBoundVarSubstitutions(e.BoundVars, forceSubstitutionOfBoundVariables: !isBindingGuard && (expr is ForallExpr || expr is ExistsExpr || expr is SetComprehension));
```
to
```
var newBoundVars = CreateBoundVarSubstitutions(e.BoundVars, forceSubstitutionOfBoundVariables: !discardSplits && (expr is ForallExpr || expr is ExistsExpr || expr is SetComprehension));
```
?
I have a hard time swallowing the second. It seems that the first one conveys the direct meaning that we are not forcing substitution because we are in a binding guard. But the second one makes a hard case: why discarding splits would result in not forcing substitution of bound variables?
We discard splits because we are in a binding guard AND we don't force substitution because we are in a binding guard. But not forcing variable substitution does not seem directly linked to discarding splits, does it?

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Substituter.cs:937]: > I have a hard time swallowing the second. It seems that the first one conveys the direct meaning that we are not forcing substitution because we are in a binding guard. But the second one makes a hard case: why discarding splits would result in not forcing substitution of bound variables?
We discard splits because we are in a binding guard AND we don't force substitution because we are in a binding guard. But not forcing variable substitution does not seem directly linked to discarding splits, does it?

I agree, but can't we just have `forceSubstitutionOfBoundVariables: false` regardless of `discardSplits` ?

Alternatively, there can be two arguments, `forceSubstitutionOfBoundVariables` and `discardSplits`.

[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyCore/Substituter.cs:940]: We should have never created splits for binding guard in the first place, that's the point of this PR.  Let me bring a fresh solution to our problem, so that I don't have to worry about the code of the substituter which sparks debate.

Let's review your two suggestions so that I explain to you my point of view
```
protected Expression SubstituteComprehensionExpr(ComprehensionExpr expr) {

  var newSplits = e switch {
    QuantifierExpr { SplitQuantifier: { } splits } => splits.Select(Substitute).ToList(),
    _ => null
  };
  ...
}
```
The problem with this suggestion is that, in the case of a binding guard, the two spitted quantifiers will be assigned variables whose scope will never be the one of the body of an if..

```
public virtual Expression Substitute(Expression expr, bool forceSubstituteOfBoundVars) {
  ...
}

protected Expression SubstituteComprehensionExpr(ComprehensionExpr expr) {

  var newSplits = e switch {
    QuantifierExpr { SplitQuantifier: { } splits } => 
      splits.Select(split => Substitute(split, forceSubstituteOfBoundVars)).ToList(),
    _ => null
  };
  ...
}
```
Here we are introducing a complexity that I don't understand. Why do we add a function to force substitute of bound vars in the global substitute function? Replacing bound variables is a purely local decision, not a variable that traverses a set of expressions. Again, it does not make sense to traverse the splits and forcing or not the substitute of bound vars in a splitted quantifier for a guard, because it did not make sense anyway that the guard be splitted if they are binding guard.

So here is my new approach. I completely reverted my changes to Substituter.cs (so that there is no debate on how to handle seemingly weirdly designed features that were out of scope of this PR).
Instead, I changed the QuantifierSplitter class, that was blindly splitting all quantifiers (except those with `{:split false}`) to ensure quantifiers just below a if guard (statement or expression) are not splitted.
That solved the trick.

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyCore/Substituter.cs]: Thanks, I like the new approach.

Any chance we can also do the first approach, in combination with your change, to make substitutor more consistent with its name and avoid future discussion about it?

```c#
protected Expression SubstituteComprehensionExpr(ComprehensionExpr expr) {

  var newSplits = e switch {
    QuantifierExpr { SplitQuantifier: { } splits } => splits.Select(Substitute).ToList(),
    _ => null
  };
  ...
}
```

**Review by keyboardDrummer — APPROVED** [review]
